solcore/absorption_calculator/tmm_core_vec.py

```diff
@@ -833,58 +879,72 @@ def inc_tmm(pol, n_list, d_list, c_list, th_0, lam_vac):
 
     # L is the transfer matrix from the i'th to (i+1)st incoherent layer, see
     # manual
+
+    # For a very opaque layer, reset T to avoid divide-by-0 and similar
+    # errors.
+    T_list[T_list < 1e-30] = 1e-30
+
     L_list = [nan]  # L_0 is not defined because 0'th layer has no beginning.
-    Ltilde = (array([[1, -R_list[1, 0]],
+    Ltilde = (array([[np.ones(len(lam_vac)), -R_list[1, 0]],
                      [R_list[0, 1],
                       T_list[1, 0] * T_list[0, 1] - R_list[1, 0] * R_list[0, 1]]])
-              / T_list[0, 1])
+              / T_list[0, 1]).transpose(2, 0, 1)
+
+    # Ltilde = Ltilde.transpose(2,0,1)
+
     for i in range(1, num_inc_layers - 1):
-        L = np.dot(
-            array([[1 / P_list[i], 0], [0, P_list[i]]]),
-            array([[1, -R_list[i + 1, i]],
+        L = np.matmul(
+            array([[1 / P_list[i], np.zeros(len(lam_vac))], [np.zeros(len(lam_vac)), P_list[i]]]).transpose(2, 0, 1),
+            array([[np.ones(len(lam_vac)), -R_list[i + 1, i]],
                    [R_list[i, i + 1],
-                    T_list[i + 1, i] * T_list[i, i + 1] - R_list[i + 1, i] * R_list[i, i + 1]]])
-        ) / T_list[i, i + 1]
+                    T_list[i + 1, i] * T_list[i, i + 1] - R_list[i + 1, i] * R_list[i, i + 1]]]).transpose(2, 0, 1)
+        ) / T_list[i, i + 1][:, None, None]
         L_list.append(L)
-        Ltilde = np.dot(Ltilde, L)
-    T = 1 / Ltilde[0, 0]
-    R = Ltilde[1, 0] / Ltilde[0, 0]
+        Ltilde = np.matmul(Ltilde, L)
+
+    #print('Ltilde', Ltilde)
+    T = 1 / Ltilde[:, 0, 0]
+    R = Ltilde[:, 1, 0] / Ltilde[:, 0, 0]
 
     # VW_list[n] = [V_n, W_n], the forward- and backward-moving intensities
     # at the beginning of the n'th incoherent layer. VW_list[0] is undefined
     # because 0'th layer has no beginning.
-    VW_list = zeros((num_inc_layers, 2))
-    VW_list[0, :] = [nan, nan]
-    VW = array([[T], [0]])
-    VW_list[-1, :] = np.transpose(VW)
+
+    VW_list = zeros((num_inc_layers, 2, len(lam_vac)))
+    VW_list[0, :, :] = nan
+    VW = array([T, zeros(len(lam_vac))])
+    VW_list[-1] = VW
+
     for i in range(num_inc_layers - 2, 0, -1):
-        VW = np.dot(L_list[i], VW)
-        VW_list[i, :] = np.transpose(VW)
+        VW = np.matmul(L_list[i], VW_list[i+1].T[:, :, None])
+        VW_list[i, :, :] = VW.transpose()
+
 
     # stackFB_list[n]=[F,B] means that F is light traveling forward towards n'th
     # stack and B is light traveling backwards towards n'th stack.
     # Reminder: inc_from_stack[i] = j means that the i'th stack comes after the
     # layer with incoherent index j.
+
     stackFB_list = []
     for stack_index, prev_inc_index in enumerate(inc_from_stack):
         if prev_inc_index == 0:  # stack starts right after semi-infinite layer.
-            F = 1
+            F = np.ones(len(lam_vac))
```

Comment:
Here, and anywhere else where you use `np.ones`, you could use [np.ones_like](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ones_like.html) instead, which does the same thing and is more explicit when you want and array identical to another one but filled with ones (or zeros, for [np.zeros_like](https://docs.scipy.org/doc/numpy/reference/generated/numpy.zeros_like.html))

Reply:
Thanks, changing it